src/metabase/native_query_analyzer.clj

```diff
@@ -37,50 +38,91 @@
+                                                  (if (seq tables)
+                                                    [:in :%lower.t/name (map normalize-name tables)]
+                                                    ;; if we don't know what tables it's from, look everywhere
+                                                    true)
+                                                  (if (seq columns)
+                                                    [:in :%lower.f/name (map normalize-name columns)]
+                                                    ;; if there are no columns, it must be a select * or similar
+                                                    false)]})))
+
+(defn- indirect-field-ids-for-query
+  "Similar to direct-field-ids-for-query, but for wildcard selects"
+  [{:keys [table-wildcards has-wildcard? tables] :as _parsed-query} db-id]
+  (let [active-fields-from-tables
+        (fn [table-names]
+          (t2/select-pks-set :model/Field (merge field-and-table-fragment
+                                                 {:where [:and
+                                                          [:= :t.db_id db-id]
+                                                          [:= true :f.active]
+                                                          (if (seq table-names)
+                                                            [:in :%lower.t/name (map normalize-name table-names)]
+                                                            ;; if no tables, nothing is selectable
+                                                            false)]})))]
+
+    (if has-wildcard?
+      ;; select * from ...
+      ;; so, get everything in all the tables
+      (active-fields-from-tables tables)
+      ;; select foo.* from ..
+      ;; limit to the named tables
+      (active-fields-from-tables table-wildcards))))
 
 (defn- field-ids-for-card
-  "Returns a list of field IDs that (may) be referenced in the given cards's query. Errs on the side of optimism:
+  "Returns a `{:direct #{...} :indirect #{...}}` map with field IDs that (may) be referenced in the given cards's query. Errs on the side of optimism:
   i.e., it may return fields that are *not* in the query, and is unlikely to fail to return fields that are in the
   query.
 
-  Returns `nil` (and logs the error) if there was a parse error."
+  Direct references are columns that are named in the query; indirect ones are from wildcards. If a field could be both direct and indirect, it will *only* show up in the `:direct` set."
   [card]
   (let [{native-query :native
-         db-id        :database} (:dataset_query card)]
-    (field-ids-for-query (:query native-query) db-id)))
+         db-id        :database} (:dataset_query card)
+        parsed-query             (mac/query->components (mac/parsed-query (:query native-query)))
+        direct-ids               (direct-field-ids-for-query parsed-query db-id)
+        indirect-ids             (set/difference
```

Comment:
Note the

```
   Direct references are columns that are named in the query; indirect ones are from wildcards. If a field could be both direct and indirect, it will *only* show up in the `:direct` set
```
   
   in the docstring

Reply:
It would be nice if we could refine our language to have different terminology for "has a wildcard reference" and "has references, but none of them are explicit". Having "indirect" switch meaning between these two is a bit annoying, and even with call-outs could trip someone up.

Reply:
> "has references, but none of them are explicit"

Hmmm I'm missing subtleness here I guess? Can you write an sql query as an example of that?

Reply:
> has a wildcard reference

This is the meaning used in `indirect-field-ids-for-query`. This includes cases like the following, where `o.product_id` is references both directly (in the join) and indirectly (in `o.*`):

```sql
select o.*, p.name as product_name from orders o join products p on o.product_id = p.id;
```

> has references, but none of them are explicit

This is the narrower meaning used in `indirect-ids`. Consider again `o.product_id` in a simpler query:

```sql
select o.* from orders;
```

Reply:
Perhaps we could call these `implicit-field-ids` and `indirect-ids` respectively, since the latter name has already made it into the db now. Thoughts @tsmacdonald?

Likewise we could make a distinction between `explicit-ids` (syntactic) and `direct-ids` (syntactic), although those would be 1-1. Still nice to have naturally pairing terms (explicit/implicit, direct/indirect).